lib/translator/common.h

```diff
@@ -23,6 +28,34 @@ void DebugLog(const char* format, ...);
 
 void SetDebugCallback(void (*callback)(const char*));
 
+void* AllocPage(uint8_t count);
+
+void SetAllocPageCallback(void* (*callback)(uint8_t));
+
+struct ScsiContext {
+  scsi_defs::LunAddress lun;
+};
+
+struct NvmeCompletionData {
+  nvme_defs::GenericQueueEntryCpl nvme_cpl[3];
+};
+
+class Translation {
+ public:
+  Translation() = default;
+  StatusCode ScsiToNvme(absl::Span<const uint8_t> scsi_cmd,
+                        const ScsiContext& scsi_context);
+  StatusCode NvmeToScsi(const NvmeCompletionData& cpl_data, void* data_in);
```

Comment:
If we're writing to `data_in`, we need *very* strong evidence that there's no danger of an arbitrarily malformed or maliciously crafted `scsi_cmd` causing us to write out of bounds. Furthermore, that evidence should be clearly cited here for the benefit of future readers.

Otherwise, we should replace this with an `absl::Span` and take care to bounds-check writes.

Reply:
At what point should we bound check writes? I'm not super familiar with memory validation best practices.

Reply:
Any time you write to memory you need to prove that the memory is properly allocated and aligned for that write to be legal. Typically the type system does most of this work for you, but with `void*` all that goes out the window.

An `absl::Span` carries dynamic length and (implicit in its template argument) static alignment information, so each write to one should first check that the range of indices being written to fall within the supplied length, unless such a check can be easily proven redundant.